src/ucp/core/ucp_listener.c

```diff
@@ -195,76 +195,134 @@ ucs_status_t ucp_listener_query(ucp_listener_h listener, ucp_listener_attr_t *at
+                                             "uct_listeners_arr");
+    if (listeners == NULL) {
+        ucs_error("Can't allocate memory for UCT listeners array");
+        return UCS_ERR_NO_MEMORY;
     }
 
-    UCP_CHECK_PARAM_NON_NULL(params->sockaddr.addr, status, return status);
+    listener->cms.listeners = listeners;
 
-    if (ucs_test_all_flags(params->field_mask,
-                           UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER |
-                           UCP_LISTENER_PARAM_FIELD_CONN_HANDLER)) {
-        ucs_error("Only one accept handler should be provided");
-        return UCS_ERR_INVALID_PARAM;
-    }
+    for (i = 0; i < num_cms; ++i) {
+        status = uct_listener_create(worker->cms[i], addr,
+                                     params->sockaddr.addrlen, &uct_params,
+                                     &listeners[listener->cms.num_listeners]);
+        if (status == UCS_ERR_INVALID_ADDR) {
+            ucs_debug("uct_listener_create failed on CM %p with address %s status %s",
+            continue;
+        } else if (status != UCS_OK) {
+            ucs_error("uct_listener_create failed on CM %p with address %s status %s",
+                      worker->cms[i],
+                      ucs_sockaddr_str(params->sockaddr.addr, addr_str,
+                                       UCS_SOCKADDR_STRING_LEN));
+            goto destroy_listeners;
+        }
 
-    UCS_ASYNC_BLOCK(&worker->async);
+        ++listener->cms.num_listeners;
 
-    listener = ucs_calloc(1, sizeof(*listener), "ucp_listener");
-    if (listener == NULL) {
-        status = UCS_ERR_NO_MEMORY;
-        goto out;
-    }
+        status = ucs_sockaddr_get_port(&addr, &port);
+        if (status != UCS_OK) {
+            goto destroy_listeners;
+        }
 
-    listener->worker = worker;
+        uct_attr.field_mask = UCT_LISTENER_ATTR_FIELD_SOCKADDR;
+        status = uct_listener_query(listeners[listener->cms.num_listeners - 1],
+                                    &uct_attr);
+        if (status != UCS_OK) {
+            goto destroy_listeners;
+        }
 
-    if (params->field_mask & UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER) {
-        UCP_CHECK_PARAM_NON_NULL(params->accept_handler.cb, status,
-                                 goto err_free_listener);
-        listener->accept_cb = params->accept_handler.cb;
-        listener->arg       = params->accept_handler.arg;
-    } else if (params->field_mask & UCP_LISTENER_PARAM_FIELD_CONN_HANDLER) {
-        UCP_CHECK_PARAM_NON_NULL(params->conn_handler.cb, status,
-                                 goto err_free_listener);
-        listener->conn_cb   = params->conn_handler.cb;
-        listener->arg       = params->conn_handler.arg;
+        status = ucs_sockaddr_get_port((struct sockaddr *)&uct_attr.sockaddr,
+                                       &listener_port);
+        if (port != listener_port) {
+            ucs_assert(port == 0);
+            ucs_sockaddr_set_port(addr, listener_port);
```

Comment:
Is it ok that we change the addr here? it's a param set by the user...

Reply:
yes, this is a copy